More tree shaking

rebornix opened the pull request:
Fixes #10307 

<!--
  If an item below does not apply to you, then go ahead and check it off as "done" and strikethrough the text, e.g.:
    - [x] ~Has unit tests & system/integration tests~
-->

-   [ ] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR).
-   [ ] Title summarizes what is changing.
-   [ ] Has a [news entry](https://github.com/Microsoft/vscode-jupyter/tree/main/news) file (remember to thank yourself!).
-   [ ] Appropriate comments and documentation strings in the code.
-   [ ] Has sufficient logging.
-   [ ] Has telemetry for feature-requests.
-   [ ] Unit tests & system/integration tests are added/updated.
-   [ ] [Test plan](https://github.com/Microsoft/vscode-jupyter/blob/main/.github/test_plan.md) is updated as appropriate.
-   [ ] [`package-lock.json`](https://github.com/Microsoft/vscode-jupyter/blob/main/package-lock.json) has been regenerated by running `npm install` (if dependencies have changed).

DonJayamanne commented on build/webpack/webpack.extension.web.config.js:
```diff
@@ -87,7 +87,7 @@ const config = {
         ]
     },
-    externals: ['vscode', 'commonjs', 'electron'], // Don't bundle these
+    externals: ['vscode', 'commonjs', 'electron', 'moment'], // Don't bundle these
```
Is it possible to do something to prevent we don't end up in a situtation where tomorrow we pull in a 3rd party package or an existing 3rd party package ends up pulling in moment, then things will not work, or it will fail only in certian cases based on where and how the 3rd party package ends up using moment.

E.g. can we write up a linter to look at package-lock.json and ensure that nothing ever pulls in moment except for jupyter. Else I feel this is a risky change (as mentioned if some other package pulls it in, then things could break sometimes)

DonJayamanne reviewed (changes requested):
Would like to ensure no other package pulls in moment, if they do, then we'd need to review that separately. Else we end up in a situation where something could break in the future.

rebornix commented:
Thanks @DonJayamanne for the suggestion, we should validate the dependencies to see if any new package we use load `moment` and fail accidentally, great catch! 

Added a gulp task to validate the lock file and added it to the pre-commit hook.

rchiodo commented on .husky/pre-commit:
```diff
@@ -15,4 +15,5 @@ if [ -z "$changed" ]; then
 fi
 
 npm run generateTelemetry
+npm run validateDependencies
```
Can we maybe do this in the linter step for the build? I'm wary of adding yet more stuff to pre-commit. Pre-commit already takes like 60 seconds for me.

rebornix replied:
@rchiodo 👍  I dislike slow commit, 60 seconds is too much. I'll move it to the linter since we always do PR.
